cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/store/DefaultMetadataStore.java

```diff
@@ -424,7 +424,8 @@ private MetadataSearchResponse search(Set<MetadataScope> scopes, String namespac
         maxEndIndex = 0;
       } else {
         startIndex = offset;
-        maxEndIndex = offset + limit;
+        // account for overflow
+        maxEndIndex = (int) Math.min(Integer.MAX_VALUE, (long) offset + (long) limit);
```

Comment:
better to do `long endIndex = offset + limit` and set maxEndIndex to `Integer.MAX_VALUE` if endIndex is greater than INT_MAX, than perform casting. also can you add a test case for this ?